eclipse-che/che-operator
# feat: implement secondsOfInactivityBeforeIdling and secondsOfRunBefor…

…eIdling CR config

Signed-off-by: David Kwon <dakwon@redhat.com>



### What does this PR do?
Part of https://github.com/eclipse/che/issues/21390.

This PR adds `spec.devEnvironments.secondsOfInactivityBeforeIdling` and `spec.devEnvironments.secondsOfRunBeforeIdling` config in the CheCluster CR:
<img width="1213" alt="image" src="https://user-images.githubusercontent.com/83611742/178339993-d5dae0b9-3400-4dd9-8256-26fff789b4f3.png">

This PR reads the above config to create a configmap named `che-idle-settings` when a user namespace is reconciled. The configmap is created so that the data can be mounted to the devworkspace containers. This PR for che-machine-exec: https://github.com/eclipse-che/che-machine-exec/pull/208 reads these environment variables. Here is an example of the generated config map:
![image](https://user-images.githubusercontent.com/83611742/178340100-a91208a5-38d7-4b0d-a163-8befcaf544b4.png)

### What issues does this PR fix or reference?

Part of https://github.com/eclipse/che/issues/21390.

### How to test this PR?



To test on an OpenShift cluster:
1. Create and install Che operator with this catalog source file:
```yaml
apiVersion:  operators.coreos.com/v1alpha1
kind:         CatalogSource
metadata:
  name:         eclipse-che-custom-catalog
  namespace:    openshift-operators
spec:
  image:       quay.io/dkwon17/eclipse-che-openshift-opm-catalog:next
  sourceType:  grpc
  updateStrategy:
    registryPoll:
      interval: 5m
```

with this che-operator image: `quay.io/dkwon17/che-operator:idle-cr`

2. Create a CheCluster CR with `spec.devEnvironments.secondsOfInactivityBeforeIdling` set to "300" and `spec.devEnvironments.secondsOfRunBeforeIdling` set to "360".

3. Open the dashboard to provision a user namespace.

4. In the user namespace verify that the `che-idle-settings` configmap in the user namespace is created, with the following data:
```
SECONDS_OF_DW_INACTIVITY_BEFORE_IDLING: "300"
SECONDS_OF_DW_RUN_BEFORE_IDLING: "360"
```
5. Edit the  `spec.devEnvironments.secondsOfInactivityBeforeIdling` and `spec.devEnvironments.secondsOfRunBeforeIdling` properties in the CheCluster CR to different values and verify that the changes exist in the `che-idle-settings` configmap.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
